include/ringmesh/basic/attributes.h

```diff
@@ -0,0 +1,1024 @@
+        * \brief Applies a permutation to the stored attributes.
+        * \details Applying a permutation to the data is equivalent
+        *  to:
+        * \code
+        * for(i=0; i<permutation.size(); i++) {
+        *    data2[i] = data[permutation[i]]
+        * }
+        * data = data2 ;
+        * \endcode
+        * But it is done in-place.
+        * \param[in] permutation the permutation.
+        *  It is temporarily changed during execution of the
+        *  function, but identical to the input on exit.
+        * \note This function uses memcpy(). If required, it
+        *  can be overloaded in derived classes.
+        */
+        void apply_permutation(
+            const std::vector< index_t >& permutation ) final
+        {
+            std::vector< T > tmp_vector;
+            tmp_vector.resize( vector_.size() );
+            for( auto p : range( permutation.size() ) )
+            {
+                tmp_vector[p] = vector_[permutation[p]];
+            }
+            vector_ = tmp_vector;
+        }
+
+    protected:
+        std::vector< T > vector_;
+    };
+
+    template < class T >
+    class VectorStore : public VectorStoreBase< T >
+    {
+    public:
+        VectorStore() = default;
+
+        Store* clone() const final
+        {
+            VectorStore< T >* result = new VectorStore< T >();
+            result->resize( this->size() );
+            result->vector_ = this->vector_;
+            return result;
+        }
+    };
+
+    template <>
+    class VectorStore< bool > : public VectorStoreBase< Byte >
+    {
+    public:
+        /**
+        * \brief Creates a new empty attribute store.
+        */
+        VectorStore() = default;
+
+        Store* clone() const final
+        {
+            VectorStore< bool >* result = new VectorStore< bool >;
+            result->resize( this->size() );
+            result->vector_ = this->vector_;
+            return result;
+        }
+        std::string element_typeid_name() const final
+        {
+            return typeid( bool ).name();
+        }
+    };
+
+    template < class T >
+    class ConstantStore : public TypedStore< T >
+    {
+    public:
+        /**
+        * \brief Creates a new empty attribute store.
+        */
+        ConstantStore()
+        {
+            this->size_ = 1;
+        }
+
+        void resize( index_t ) final
+        {
+            Logger::warn( "Trying to resize constant attribute" );
+        }
+
+        const void* data() const final
+        {
+            return &constant_;
+        }
+        void* data() final
+        {
+            return &constant_;
+        }
+        const void* element( index_t element ) const final
+        {
+            ringmesh_unused( element );
+            return &this->constant_;
+        }
+        void* element( index_t element ) final
+        {
+            ringmesh_unused( element );
+            return &this->constant_;
+        }
+        void clear() final
+        {

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Add a nested comment explaining why this method is empty, or complete the implementation. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS1186)